### DISCUSSION TITLE
How to restrict API write access to tables in favor of functions?

### DISCUSSION THREAD

**Discussion by stefan-girlich** [supabase/supabase#22434] [Questions] [answered]
I am planning to use PostgreSQL functions for write access to the database since they allow for using transactions and enforcing validation rules, e.g. _Insertion in table "groups" is only allowed when the new group is at least referenced by three rows in table "members"_.
But an authenticated user could still bypass these functions and violate these rules, e.g. by inserting a row into "groups" without linking the corresponding rows from "members".

Is there a way to...
- enforce authenticated write access through functions
- disable direct write access via API/SDK
- allow read access via API/SDK?

**Comment by stefan-girlich** [comment]
The documentation contains a possible solution:
https://supabase.com/docs/guides/api/securing-your-api#enforce-additional-rules-on-each-request

**Comment by GaryAustin1** [comment, marked as answer]
The easiest way is to have RLS enabled on the tables with no insert/update/delete policies.  Have a select policy.
Then make your function security definer type so it bypasses RLS.

**Reply by parkernilson** [reply]
I currently do it this way as well, however does that present security concerns because you are exposing SECURITY DEFINER functions to the public?

**Reply by GaryAustin1** [reply]
https://supabase.com/docs/guides/database/functions#security-definer-vs-invoker
Just limit the search path.
